[BUG] SchemaError on a valid OpenAPI 3.1.0 schema

### Checklist

- [X] I checked the [FAQ section](https://schemathesis.readthedocs.io/en/stable/faq/) of the documentation
- [X] I looked for similar issues in the [issue tracker](https://github.com/schemathesis/schemathesis/issues)
- [X] I am using the latest version of Schemathesis

### Describe the bug

I'm using Schemathesis 4.2.0 & PyTest 8.4.2. I'm trying to test my API with an OpenAPI 3.1.0 schema (see bellow) generated by FastAPI 0.118.2. I've 30+ routes that Schemathesis which all doesn't give errors except for the route `/location/all`. During testing it gives the following error: `jsonschema.exceptions.SchemaError: ['number', 'integer', 'integer'] is not valid under any of the given schemas` (see complete stack trace below).

The original schema doesn't have multiple data types set (see simplified schema below). Looking at the stack trace/exception below it seems that during canonicalisation Schemathesis or one of the other used libraries ends up merging multiple schema fragments producing `["number", "integer", "integer"]`. As far as I understand/my online research tells me this should be no problem with OpenAPI 3.1.0 (unlike 3.0.x). Maybe this has to do with the minimum & maximum (since this is the only route with query parameters with minimum & maximum constraints.

Complete stack trace:
```
    @wraps(test_function)
>   def test_wrapper(*args: Any, **kwargs: Any) -> Any:
               ^^^

../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/schemathesis/generation/hypothesis/builder.py:224: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/schemathesis/specs/openapi/_hypothesis.py:101: in openapi_cases
    query_ = generate_parameter(
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/schemathesis/specs/openapi/_hypothesis.py:302: in generate_parameter
    value = get_parameters_value(explicit, location, draw, operation, ctx, hooks, strategy_factory, generation_config)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/schemathesis/specs/openapi/_hypothesis.py:246: in get_parameters_value
    return draw(strategy)
           ^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis/internal/conjecture/data.py:1212: in draw
    return unwrapped.do_draw(self)
           ^^^^^^^^^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis/strategies/_internal/strategies.py:1014: in do_draw
    x = data.draw(self.mapped_strategy)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis/internal/conjecture/data.py:1212: in draw
    return unwrapped.do_draw(self)
           ^^^^^^^^^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis/strategies/_internal/strategies.py:1191: in do_draw
    result = self.do_filtered_draw(data)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis/strategies/_internal/strategies.py:1200: in do_filtered_draw
    value = data.draw(self.filtered_strategy)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis/internal/conjecture/data.py:1212: in draw
    return unwrapped.do_draw(self)
           ^^^^^^^^^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis/strategies/_internal/flatmapped.py:60: in do_draw
    return data.draw(expanded)
           ^^^^^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis/internal/conjecture/data.py:1212: in draw
    return unwrapped.do_draw(self)
           ^^^^^^^^^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis/strategies/_internal/strategies.py:1191: in do_draw
    result = self.do_filtered_draw(data)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis/strategies/_internal/strategies.py:1200: in do_filtered_draw
    value = data.draw(self.filtered_strategy)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis/internal/conjecture/data.py:1212: in draw
    return unwrapped.do_draw(self)
           ^^^^^^^^^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis/strategies/_internal/core.py:1071: in do_draw
    obj = self.target(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis_jsonschema/_from_schema.py:146: in error_raiser
    raise error
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis_jsonschema/_from_schema.py:137: in from_schema
    return __from_schema(
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis_jsonschema/_from_schema.py:208: in __from_schema
    schema = canonicalish(schema)
             ^^^^^^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis_jsonschema/_canonicalise.py:515: in canonicalish
    m = merged([not_, {**schema, "type": type_}])
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis_jsonschema/_canonicalise.py:709: in merged
    m = merged([out_combined, s_combined])
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis_jsonschema/_canonicalise.py:631: in merged
    schemas = sorted((canonicalish(s) for s in schemas), key=upper_bound_instances)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis_jsonschema/_canonicalise.py:631: in <genexpr>
    schemas = sorted((canonicalish(s) for s in schemas), key=upper_bound_instances)
                      ^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis_jsonschema/_canonicalise.py:515: in canonicalish
    m = merged([not_, {**schema, "type": type_}])
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis_jsonschema/_canonicalise.py:853: in merged
    _get_validator_class(out)
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis_jsonschema/_canonicalise.py:95: in _get_validator_class
    return __get_validator_class(CacheableSchema(schema))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/hypothesis_jsonschema/_canonicalise.py:108: in __get_validator_class
    jsonschema.Draft4Validator.check_schema(schema)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cls = <class 'jsonschema.validators.Draft4Validator'>
schema = {'maximum': 90, 'minimum': -90, 'title': 'Max Latitude', 'type': ['number', 'integer', 'integer']}
format_checker = <FormatChecker checkers=['date-time', 'email', 'hostname', 'idn-email', 'ipv4', 'ipv6', 'regex', 'uri']>

    @classmethod
    def check_schema(cls, schema, format_checker=_UNSET):
        Validator = validator_for(cls.META_SCHEMA, default=cls)
        if format_checker is _UNSET:
            format_checker = Validator.FORMAT_CHECKER
        validator = Validator(
            schema=cls.META_SCHEMA,
            format_checker=format_checker,
        )
        for error in validator.iter_errors(schema):
>           raise exceptions.SchemaError.create_from(error)
E           jsonschema.exceptions.SchemaError: ['number', 'integer', 'integer'] is not valid under any of the given schemas
E           
E           Failed validating 'anyOf' in metaschema['properties']['type']:
E               {'anyOf': [{'$ref': '#/definitions/simpleTypes'},
E                          {'type': 'array',
E                           'items': {'$ref': '#/definitions/simpleTypes'},
E                           'minItems': 1,
E                           'uniqueItems': True}]}
E           
E           On schema['type']:
E               ['number', 'integer', 'integer']
E           while generating 'case' from one_of(openapi_cases(operation=APIOperation(path='/location/all',
E            method='get',
E            definition=,
E            schema=,
E            responses=OpenApiResponses(_inner={'200': OpenApiResponse(status_code='200',
E               definition={'description': 'Successful Response',
E                'content': {'application/json': {'schema': {'type': 'array',
E                   'items': {'$ref': '#/components/schemas/Location'},
E                   'title': 'Response Locations Get Location All Get'}}}},
E               resolver=<schemathesis.specs.openapi.references.ReferenceResolver object at 0x10d3ae120>,
E               scope='/openapi.json',
E               adapter=<module 'schemathesis.specs.openapi.adapter.v3_1' from '/Users/prive/Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/schemathesis/specs/openapi/adapter/v3_1.py'>),
E              '422': OpenApiResponse(status_code='422',
E               definition={'description': 'Validation Error',
E                'content': {'application/json': {'schema': {'$ref': '#/components/schemas/HTTPValidationError'}}}},
E               resolver=<schemathesis.specs.openapi.references.ReferenceResolver object at 0x10d3ae120>,
E               scope='/openapi.json',
E               adapter=<module 'schemathesis.specs.openapi.adapter.v3_1' from '/Users/prive/Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/schemathesis/specs/openapi/adapter/v3_1.py'>)},
E             resolver=<schemathesis.specs.openapi.references.ReferenceResolver object at 0x10d3ae120>,
E             scope='/openapi.json',
E             adapter=<module 'schemathesis.specs.openapi.adapter.v3_1' from '/Users/prive/Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/schemathesis/specs/openapi/adapter/v3_1.py'>),
E            security=OpenApiSecurityParameters(_parameters=[]),
E            label='GET /location/all',
E            app=<fastapi.applications.FastAPI object at 0x10d40d810>,
E            base_url='/',
E            path_parameters=,
E            headers=,
E            cookies=,
E            query=,
E            body=)), openapi_cases(operation=APIOperation(path='/location/all',
E            method='get',
E            definition=,
E            schema=,
E            responses=OpenApiResponses(_inner={'200': OpenApiResponse(status_code='200',
E               definition={'description': 'Successful Response',
E                'content': {'application/json': {'schema': {'type': 'array',
E                   'items': {'$ref': '#/components/schemas/Location'},
E                   'title': 'Response Locations Get Location All Get'}}}},
E               resolver=<schemathesis.specs.openapi.references.ReferenceResolver object at 0x10d3ae120>,
E               scope='/openapi.json',
E               adapter=<module 'schemathesis.specs.openapi.adapter.v3_1' from '/Users/prive/Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/schemathesis/specs/openapi/adapter/v3_1.py'>),
E              '422': OpenApiResponse(status_code='422',
E               definition={'description': 'Validation Error',
E                'content': {'application/json': {'schema': {'$ref': '#/components/schemas/HTTPValidationError'}}}},
E               resolver=<schemathesis.specs.openapi.references.ReferenceResolver object at 0x10d3ae120>,
E               scope='/openapi.json',
E               adapter=<module 'schemathesis.specs.openapi.adapter.v3_1' from '/Users/prive/Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/schemathesis/specs/openapi/adapter/v3_1.py'>)},
E             resolver=<schemathesis.specs.openapi.references.ReferenceResolver object at 0x10d3ae120>,
E             scope='/openapi.json',
E             adapter=<module 'schemathesis.specs.openapi.adapter.v3_1' from '/Users/prive/Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/schemathesis/specs/openapi/adapter/v3_1.py'>),
E            security=OpenApiSecurityParameters(_parameters=[]),
E            label='GET /location/all',
E            app=<fastapi.applications.FastAPI object at 0x10d40d810>,
E            base_url='/',
E            path_parameters=,
E            headers=,
E            cookies=,
E            query=,
E            body=), generation_mode=<GenerationMode.NEGATIVE: 'negative'>))

../../../Library/Caches/pypoetry/virtualenvs/backend-CK-nuquT-py3.13/lib/python3.13/site-packages/jsonschema/validators.py:317: SchemaError
```

### To Reproduce

🚨 **Mandatory** 🚨: Steps to reproduce the behavior:

Test file content:
```
import pytest
import schemathesis
from schemathesis.specs.openapi.schemas import BaseOpenAPISchema


@pytest.fixture
def api_schema() -> BaseOpenAPISchema:
    return schemathesis.openapi.from_path('openapi.json')


schema = schemathesis.pytest.from_fixture('api_schema').include(path="/location/all")


@schema.parametrize()
def test_api(case):
    case.call_and_validate(base_url='http://localhost:8000')
```
- Webserver implementation not needed as long as webserver ia s running on the specified port.
- Use the minimal API schema below and store the contents in a file called openapi.json.
- Run the test with PyTest.

Minimal API schema causing this issue:

```yaml
{
    "openapi": "3.1.0",
    "paths": {
        "/location": {
            "get": {
                "tags": [
                    "Locations"
                ],
                "summary": "Locations Get",
                "operationId": "locations_get_location_all_get",
                "parameters": [
                    {
                        "name": "min_longitude",
                        "in": "query",
                        "required": true,
                        "schema": {
                            "type": "number",
                            "maximum": 180,
                            "minimum": -180,
                            "title": "Min Longitude"
                        }
                    },
                    {
                        "name": "max_longitude",
                        "in": "query",
                        "required": true,
                        "schema": {
                            "type": "number",
                            "maximum": 180,
                            "minimum": -180,
                            "title": "Max Longitude"
                        }
                    },
                    {
                        "name": "min_latitude",
                        "in": "query",
                        "required": true,
                        "schema": {
                            "type": "number",
                            "maximum": 90,
                            "minimum": -90,
                            "title": "Min Latitude"
                        }
                    },
                    {
                        "name": "max_latitude",
                        "in": "query",
                        "required": true,
                        "schema": {
                            "type": "number",
                            "maximum": 90,
                            "minimum": -90,
                            "title": "Max Latitude"
                        }
                    },
                    {
                        "name": "start_time",
                        "in": "query",
                        "required": true,
                        "schema": {
                            "type": "string",
                            "format": "date-time",
                            "title": "Start Time"
                        }
                    },
                    {
                        "name": "end_time",
                        "in": "query",
                        "required": true,
                        "schema": {
                            "type": "string",
                            "format": "date-time",
                            "title": "End Time"
                        }
                    }
                ],
                "responses": {
                    "200": {
                        "description": "Successful Response",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "array",
                                    "items": {
                                        "$ref": "#/components/schemas/Location"
                                    },
                                    "title": "Response Locations Get Location All Get"
                                }
                            }
                        }
                    },
                    "422": {
                        "description": "Validation Error",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/HTTPValidationError"
                                }
                            }
                        }
                    }
                }
            }
        }
    },
    "components": {
        "schemas": {
            "HTTPValidationError": {
                "properties": {
                    "detail": {
                        "items": {
                            "$ref": "#/components/schemas/ValidationError"
                        },
                        "type": "array",
                        "title": "Detail"
                    }
                },
                "type": "object",
                "title": "HTTPValidationError"
            },
            "Location": {
                "properties": {
                    "uuids": {
                        "items": {
                            "type": "string",
                            "format": "uuid"
                        },
                        "type": "array",
                        "title": "Uuids",
                        "default": []
                    },
                    "latitude": {
                        "type": "number",
                        "exclusiveMaximum": 53.85,
                        "exclusiveMinimum": 50.45,
                        "title": "Latitude"
                    },
                    "longitude": {
                        "type": "number",
                        "exclusiveMaximum": 7.48,
                        "exclusiveMinimum": 3.11,
                        "title": "Longitude"
                    }
                },
                "type": "object",
                "required": [
                    "latitude",
                    "longitude"
                ],
                "title": "Location"
            },
            "ValidationError": {
                "properties": {
                    "loc": {
                        "items": {
                            "anyOf": [
                                {
                                    "type": "string"
                                },
                                {
                                    "type": "integer"
                                }
                            ]
                        },
                        "type": "array",
                        "title": "Location"
                    },
                    "msg": {
                        "type": "string",
                        "title": "Message"
                    },
                    "type": {
                        "type": "string",
                        "title": "Error Type"
                    }
                },
                "type": "object",
                "required": [
                    "loc",
                    "msg",
                    "type"
                ],
                "title": "ValidationError"
            }
        }
    }
}
```

### Expected behavior

No 'SchemaError' with the specified OpenAPI schema.

### Environment

```
- OS: MacOS 15.7
- Python version: 3.13.7
- PyTest 8.4.2
- FastAPI 0.118.2
- Schemathesis version: 4.2.0
- Spec version: Open API 3.1.0
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,10 @@
 
 ## [Unreleased](https://github.com/schemathesis/schemathesis/compare/v4.2.0...HEAD) - TBD
 
+### :bug: Fixed
+
+- False positive schema error message due to incorrect schema type detection. [#3149](https://github.com/schemathesis/schemathesis/issues/3149)
+
 ### :racing_car: Performance
 
 - Up to ~20% faster VCR cassettes recording.

--- a/src/schemathesis/core/jsonschema/types.py
+++ b/src/schemathesis/core/jsonschema/types.py
@@ -17,7 +17,7 @@ def get_type(schema: JsonSchema, *, _check_type: bool = False) -> list[str]:
         return [ty]
     if ty is ANY_TYPE:
         return list(ty)
-    return list(ty)
+    return [t for t in ALL_TYPES if t in ty]
 
 
 def _get_type(schema: JsonSchema) -> list[str]:

--- a/test/core/jsonschema/test_types.py
+++ b/test/core/jsonschema/test_types.py
@@ -10,7 +10,8 @@ from schemathesis.core.jsonschema.types import ANY_TYPE, get_type, to_json_type_
         (False, ANY_TYPE),
         ({}, ANY_TYPE),
         ({"type": "string"}, ["string"]),
-        ({"type": ["string", "null"]}, ["string", "null"]),
+        ({"type": ["string", "null"]}, ["null", "string"]),
+        ({"type": ["integer", "integer"]}, ["integer"]),
     ],
 )
 def test_get_type(schema, expected):

--- a/test/test_parameters.py
+++ b/test/test_parameters.py
@@ -788,3 +788,28 @@ def test_request_body_with_boolean_true_schema(ctx, cli, openapi3_base_url, snap
         cli.run(str(schema_path), "--max-examples=1", f"--url={openapi3_base_url}", "--checks=not_a_server_error")
         == snapshot_cli
     )
+
+
+def test_parameter_type_detection(ctx, cli, openapi3_base_url, snapshot_cli):
+    # See GH-3149
+    schema_path = ctx.openapi.write_schema(
+        {
+            "/success": {
+                "get": {
+                    "parameters": [
+                        {
+                            "name": "longitude",
+                            "in": "query",
+                            "schema": {
+                                "maximum": 180,
+                            },
+                        }
+                    ]
+                }
+            }
+        }
+    )
+    assert (
+        cli.run(str(schema_path), f"--url={openapi3_base_url}", "--checks=not_a_server_error", "--max-examples=5")
+        == snapshot_cli
+    )

